lib/logstash/filters/util/lookup.rb

```diff
@@ -0,0 +1,156 @@
+    end
+
+    class Getfier
+      def initialize(param)
+        @param = param
+      end
+
+      def fetch(event)
+        event.get(@param)
+      end
+    end
+
+    attr_reader :target, :query, :parameters
+
+    def self.validate(array_of_options)
+      errors = []
+      array_of_options.each_with_index do |options, i|
+        instance = new(options, {}, Object.new, "lookup-#{i.next}")
+        unless instance.valid?
+          errors << instance.formatted_errors
+        end
+      end
+      return nil if errors.empty?
+      errors.join("; ")
+    end
+
+    def initialize(options, globals, logger, default_id)
+      @target = options["target"]
+      @id = target || default_id
+      @options = options
+      @globals = globals
+      @logger = logger
+      @valid = false
+      @option_errors = []
+      @default_array = nil
+      parse_options
+    end
+
+    def valid?
+      @valid
+    end
+
+    def formatted_errors
+      @option_errors.join(", ")
+    end
+
+    def enhance(local, event)
+      result = fetch(local, event) # should return a LookupResult
+
+      if result.failed?
+        tag_failure(event)
+      end
+
+      if result.empty? && @use_default
+        tag_default(event)
+        process_event(event, @default_array)
+        return
+      end
+      process_event(event, result.payload)
+    end
+
+    # ------------------
+    private
+
+
+    def tag_failure(event)
+      @tag_on_failure.each do |tag|
+        event.tag(tag)
+      end
+    end
+
+    def tag_default(event)
+      @tag_on_default_use.each do |tag|
+        event.tag(tag)
+      end
+    end
+
+    def fetch(local, event)
+      result = LookupResult.new()
+      params = prepare_parameters_from_event(event)
+      begin
+        @logger.debug? && @logger.debug("Executing JDBC query", :statement => query, :parameters => params)
+        local.fetch(query, params).each do |row|
+          result.push row.inject({}){|hash,(k,v)| hash[k.to_s] = v; hash} #Stringify row keys
+        end
+      rescue ::Sequel::Error => e
+        # all sequel errors are a subclass of this, let all other standard or runtime errors bubble up
+        result.failed!
+        @logger.warn? && @logger.warn("Exception when executing JDBC query", :exception => e)
+      end
+      # if either of: no records or a Sequel exception occurs the payload is
+      # empty and the default can be substituted later.
+      result
+    end
+
+    def process_event(event, value)
+      # use deep clone here so other filter function don't taint the payload by reference
+      event.set(@target, ::LogStash::Util.deep_clone(value))
+    end
+
+    def prepare_parameters_from_event(event)
+      @symbol_parameters.inject({}) do |hash,(k,v)|
+        value = v.fetch(event)
+        hash[k] = value.is_a?(::LogStash::Timestamp) ? value.time : value
+        hash
+      end
+    end
+
+    def sprintf_or_get(v)
+      v.include?("%{") ? Sprintfier.new(v) : Getfier.new(v)
+    end
```

Comment:
OK, make sense, but maybe we can make the regular expression a bit more complete and use the one defined in the `StringInterpolation` class https://github.com/elastic/logstash/blob/6fb8096a54b3cf3b3c1f2ad77ab9d167b0e864cb/logstash-core/src/main/java/org/logstash/StringInterpolation.java#L11